HELP.md

```diff
@@ -0,0 +1,717 @@
+# Contract best practices
+
+## Main structure and persistent collections
+
+The main contract structure is marked with `#[near_bindgen]`. It has to be serializable and deserializable with [Borsh](https://borsh.io).
+
+```rust
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct Contract {
+    pub data: String,
+    pub owner_id: AccountId,
+    pub value: u128,
+}
+```
+
+Every time an external method is called, the entire structure has to be deserialized.
+The serialized contract data is stored in the persistent storage under the key `STATE`.
+
+Change methods (see below) are serializing the main contract structure at the end and stores the new value into the storage.
+
+Persistent collection helps store extra data in the persistent storage outside of main structure.
+NEAR SDK provides the following collections:
+- `Vector` - An iterable implementation of vector.
+- `LookupMap` - An non-iterable implementation of a map.
+- `LookupSet` - An non-iterable implementation of a set.
+- `UnorderedMap` - An iterable implementation of a map.
+- `UnorderedSet` - An iterable implementation of a set.
+- `TreeMap` - An iterable sorted map based on AVL-tree
+- `LazyOption` - An `Option` for a single value.
+
+Every instance of a persistent collection requires a unique storage prefix.
+The prefix is used to generate internal keys to store data in the persistent storage.
+These internal keys has to be unique to avoid collisions (even with key `STATE`).
+
+## Generating unique prefixes for persistent collections
+
+When a contract gets complicated, there may be multiple different
+collections that may not be all part of the main structure, but instead be part of sub-structure or nested collections.
+They all need to have unique prefixes.
+
+We can introduce an `enum` for tracking storage prefixes and keys.
+And then use borsh serialization to construct a unique prefix for every collection.
+It's as efficient as manually constructing them, because Borsh enum only takes one byte.
+
+```rust
+#[derive(BorshSerialize)]
+pub enum StorageKeys {
+    Accounts,
+    SubAccount { account_hash: Vec<u8> },
+    Tokens,
+    Metadata,
+}
+
+#[near_bindgen]
+impl Contract {
+    #[init]
+    pub fn new() -> Self {
+        Self {
+            accounts: UnorderedMap::new(StorageKeys::Accounts.try_to_vec().unwrap()),
+            tokens: LookupMap::new(StorageKeys::Tokens.try_to_vec().unwrap()),
+            metadata: LazyOption::new(StorageKeys::Metadata.try_to_vec().unwrap()),
+        }
+    }
+    
+    fn get_tokens(&self, account_id: &AccountId) -> UnorderedSet<String> {
+        let tokens = self.accounts.get(account_id).unwrap_or_else(|| {
+            UnorderedSet::new(
+                StorageKeys::SubAccount { account_hash: env::sha256(account_id.as_bytes()) }
+                    .try_to_vec()
+                    .unwrap(),
```

Comment:
is there a way to avoid all the `.try_to_vec().unwrap()`? It really clutters up a great interface.

Reply:
we can add `to_key` impl that does this:
```rust
impl StorageKeys {
  pub fn to_vec(&self) -> Vec<u8> {
    self.try_to_vec().unwrap()
  }
}
```

Reply:
We can also change `UnorderedSet::new` to take `IntoVec` and then it may simplify the code.

Reply:
Would we have to do that with every kind of collection? Is it worth it? What would this look like, then? This?

```rust
StorageKeys::SubAccount { account_hash: env::sha256(account_id.as_bytes()) }
    into(),
```